File: src/main/java/io/github/shaksternano/borgar/media/io/MediaWriters.java
package io.github.shaksternano.borgar.media.io;

import io.github.shaksternano.borgar.media.io.writer.*;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class MediaWriters {

    private static final Map<String, MediaWriterFactory> writerFactories = new HashMap<>();

    public static MediaWriter createWriter(
        File output,
        String outputFormat,
        int loopCount,
        int audioChannels,
        int audioSampleRate,
        int audioBitrate,
        long maxFileSize,
        long maxDuration
    ) throws IOException {
        outputFormat = outputFormat.toLowerCase();
        var factory = writerFactories.getOrDefault(outputFormat, (
            output1,
            outputFormat1,
            loopCount1,
            audioChannels1,
            audioSampleRate1,
            audioBitrate1,
            maxFileSize1,
            maxDuration1
        ) -> new FFmpegVideoWriter(output1, outputFormat1, audioChannels1, audioSampleRate1, audioBitrate1, maxFileSize1, maxDuration1));
        return factory.createWriter(
            output,
            outputFormat,
            loopCount,
            audioChannels,
            audioSampleRate,
            audioBitrate,
            maxFileSize,
            maxDuration
        );
    }

    private static void registerWriterFactory(MediaWriterFactory factory, String... formats) {
        for (var format : formats) {
            writerFactories.putIfAbsent(format.toLowerCase(), factory);
        }
    }

    static {
        registerWriterFactory(
            (
                output,
                outputFormat,
                loopCount,
                audioChannels,
                audioSampleRate,
                audioBitrate,
                maxFileSize,
                maxDuration
            ) -> new ScrimageGifWriter(output, loopCount),
            "gif"
        );
        registerWriterFactory(
            (
                output,
                outputFormat,
                loopCount,
                audioChannels,
                audioSampleRate,
                audioBitrate,
                maxFileSize,
                maxDuration
            ) -> new JavaxImageWriter(output, outputFormat),
            "bmp",
            "jpeg",
            "jpg",
            "wbmp",
            "png",
            "gif",
            "tif",
            "tiff"
        );
        registerWriterFactory(
            (
                output,
                outputFormat,
                loopCount,
                audioChannels,
                audioSampleRate,
                audioBitrate,
                maxFileSize,
                maxDuration
            ) -> new Image4jIcoWriter(output),
            "ico"
        );
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/MediaWriters.kt
package io.github.shaksternano.borgar.core.media

import io.github.shaksternano.borgar.core.collect.putAllKeys
import io.github.shaksternano.borgar.core.media.writer.*
import java.nio.file.Path
import kotlin.time.Duration

private val writerFactories: Map<String, MediaWriterFactory> = buildMap {
    registerFactory(JavaxImageWriter.Factory)
    registerFactory(ScrimageGifWriter.Factory)
    registerFactory(Image4jIcoWriter.Factory)
}

private fun MutableMap<String, MediaWriterFactory>.registerFactory(
    factory: MediaWriterFactory,
) = putAllKeys(
    factory.supportedFormats,
    factory,
)

fun isWriterFormatSupported(format: String): Boolean =
    writerFactories.containsKey(format)

suspend fun createWriter(
    output: Path,
    outputFormat: String,
    loopCount: Int,
    audioChannels: Int,
    audioSampleRate: Int,
    audioBitrate: Int,
    maxFileSize: Long,
    maxDuration: Duration,
): MediaWriter {
    val factory = writerFactories.getOrDefault(outputFormat, FFmpegVideoWriter.Factory)
    return factory.create(
        output,
        outputFormat,
        loopCount,
        audioChannels,
        audioSampleRate,
        audioBitrate,
        maxFileSize,
        maxDuration,
    )
}
